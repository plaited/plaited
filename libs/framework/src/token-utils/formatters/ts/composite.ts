import { camelCase } from '@plaited/utils'
import { Formatter, CompositeToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
