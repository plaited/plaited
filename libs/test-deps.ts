// -- std --
export {
  assert,
  assertEquals,
  assertFalse,
} from 'https://deno.land/std@0.176.0/testing/asserts.ts'
export { assertSnapshot } from 'https://deno.land/std@0.176.0/testing/snapshot.ts'
export {
  afterEach,
  beforeEach,
  describe,
  it,
} from 'https://deno.land/std@0.176.0/testing/bdd.ts'
export { resolve } from 'https://deno.land/std@0.175.0/path/mod.ts'
// -- sinon --
export { default as sinon } from 'https://esm.sh/sinon@15.0.1'

// -- beautify --
export { default as beautify } from 'https://esm.sh/beautify@0.0.8'
