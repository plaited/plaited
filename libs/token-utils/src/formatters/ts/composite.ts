import { camelCase } from '@plaited/utils'
import { Formatter, AliasValue, FlexToken, GridToken, TypographyToken } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar } from '../ts-utils.js'

type CompositeToken = FlexToken | GridToken | TypographyToken

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value as AliasValue, allTokens)}`
  }
  const toRet = Object.entries($value).map(([key, val]) => `  ${key}: ${resolveTSVar(val, allTokens)},`)
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
