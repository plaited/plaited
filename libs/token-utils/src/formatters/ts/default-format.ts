import { camelCase } from '@plaited/utils'
import { Formatter } from '../../types.js'
import { hasAlias } from '../has-alias.js'
import { isValidAlias, getCssVar } from '../ts-utils.js'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
