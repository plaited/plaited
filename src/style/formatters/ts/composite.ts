import type { CompositeToken } from '../../token.types.js'
import type { Formatter } from '../formatters.types.js'
import { camelCase } from '../../../utils/case.js'
import { hasAlias } from '../has-alias.js'
import { resolveTSVar, isValidAlias } from '../ts-utils.js'

export const composite: Formatter<CompositeToken> = (token, { tokenPath, allTokens }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  if (hasAlias($value)) {
    return `export const ${camelCase(tokenPath.join(' '))} = ${resolveTSVar($value, allTokens)}`
  }
  const toRet: string[] = []
  for (const key in $value) {
    const val = $value[key]
    if (isValidAlias(val, allTokens)) {
      toRet.push(`  ${key}: ${resolveTSVar(val, allTokens)},`)
    }
  }
  return [`export const ${camelCase(tokenPath.join(' '))} = {`, ...toRet, '}'].join('\n')
}
