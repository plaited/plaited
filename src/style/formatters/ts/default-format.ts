import type { Formatter } from '../formatters.types.ts'
import { camelCase } from '../../../utils/case.ts'
import { hasAlias } from '../has-alias.ts'
import { isValidAlias, getCssVar } from '../ts-utils.ts'

export const defaultFormat: Formatter = (token, { allTokens, tokenPath }) => {
  const { $value } = token
  if (hasAlias($value) && !isValidAlias($value, allTokens)) {
    return ''
  }
  return `export const ${camelCase(tokenPath.join(' '))} = ${getCssVar(tokenPath)}`
}
