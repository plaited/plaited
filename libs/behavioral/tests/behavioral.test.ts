import { assertEquals, assertSnapshot } from '../../test-deps.ts'
import { bProgram, bThread, loop, sync } from '../mod.ts'

const expectedFeedback = [
  'Add hot',
  'Add cold',
  'Add hot',
  'Add cold',
  'Add hot',
  'Add cold',
]

const threads = {
  addHot: bThread(
    sync<{ value: string }>({
      waitFor: [{
        cb({ detail }) {
          return detail?.value === 'start'
        },
      }, { event: 'say_cold' }],
    }),
    sync({
      request: { event: 'hot' },
    }),
    sync({
      request: { event: 'hot' },
    }),
    sync({
      request: { event: 'hot' },
    }),
  ),
  addCold: bThread(
    sync({ waitFor: { event: 'start' } }),
    sync({
      request: [{ event: 'cold' }],
    }),
    sync({
      request: [{ event: 'cold' }],
    }),
    sync({
      request: [{ event: 'cold' }],
    }),
  ),
  mixHotCold: loop(
    bThread(
      sync({
        waitFor: { event: 'hot' },
        block: { event: 'cold' },
      }),
      sync({
        waitFor: [{ event: 'cold' }],
        block: [{ event: 'hot' }],
      }),
    ),
  ),
}
const getActions = (arr: string[]) => ({
  cold() {
    arr.push('Add cold')
  },
  hot() {
    arr.push('Add hot')
  },
})
Deno.test('bProgram: priority queue', (t) => {
  const actualFeedback: string[] = []
  const logs: unknown[] = []
  const { trigger, feedback, log, add } = bProgram({
    dev: true,
  })
  add(threads)
  feedback(getActions(actualFeedback))
  log((msg) => {
    logs.push(msg)
  })
  trigger({
    event: 'start',
    detail: { value: 'start' },
  })
  assertEquals(
    actualFeedback,
    expectedFeedback,
    `priority selection feedback`,
  )
  assertSnapshot(t, logs, `priority selection feedback`)
})
Deno.test('bProgram: randomized priority queue', (t) => {
  const actualFeedback: string[] = []
  const logs: unknown[] = []
  const { trigger, feedback, log, add } = bProgram({
    strategy: 'randomized',
    dev: true,
  })
  add(threads)
  feedback(getActions(actualFeedback))
  log((msg) => {
    logs.push(msg)
  })
  trigger({
    event: 'start',
    detail: { value: 'start' },
  })
  assertEquals(
    actualFeedback,
    expectedFeedback,
    `randomized priority selection feedback`,
  )
  assertSnapshot(t, logs, `randomized priority selection log`)
})
Deno.test('bProgram: chaos selection', (t) => {
  const actualFeedback: string[] = []
  const logs: unknown[] = []
  const { trigger, feedback, log, add } = bProgram({
    strategy: 'chaos',
    dev: true,
  })
  add(threads)
  feedback(getActions(actualFeedback))
  log((msg) => {
    logs.push(msg)
  })
  trigger({
    event: 'start',
    detail: { value: 'start' },
  })
  assertEquals(actualFeedback, expectedFeedback, `chaos selection feedback`)
  assertSnapshot(t, logs, `chaos selection log`)
})
