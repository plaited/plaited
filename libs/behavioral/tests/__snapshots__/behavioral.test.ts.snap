export const snapshot = {};

snapshot[`bProgram: priority queue 1`] = `
[
  {
    data: {
      detail: {
        value: "start",
      },
      event: "start",
    },
    type: "trigger",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          waitFor: [
            {
              cb: [Function: cb],
            },
            {
              event: "say_cold",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          waitFor: [
            {
              event: "start",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "start",
          priority: 0,
          request: [
            {
              detail: {
                value: "start",
              },
              event: "start",
            },
          ],
          waitFor: [
            {
              cb: [Function: cb],
              event: "",
            },
          ],
        },
      ],
      selectedEvent: {
        detail: {
          value: "start",
        },
        event: "start",
        priority: 0,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      strategy: "priority",
    },
    type: "end",
  },
]
`;

snapshot[`bProgram: randomized priority queue 1`] = `
[
  {
    data: {
      detail: {
        value: "start",
      },
      event: "start",
    },
    type: "trigger",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          waitFor: [
            {
              cb: [Function: cb],
            },
            {
              event: "say_cold",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          waitFor: [
            {
              event: "start",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "start",
          priority: 0,
          request: [
            {
              detail: {
                value: "start",
              },
              event: "start",
            },
          ],
          waitFor: [
            {
              cb: [Function: cb],
              event: "",
            },
          ],
        },
      ],
      selectedEvent: {
        detail: {
          value: "start",
        },
        event: "start",
        priority: 0,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      strategy: "randomized",
    },
    type: "end",
  },
]
`;

snapshot[`bProgram: chaos selection 1`] = `
[
  {
    data: {
      detail: {
        value: "start",
      },
      event: "start",
    },
    type: "trigger",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          waitFor: [
            {
              cb: [Function: cb],
            },
            {
              event: "say_cold",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          waitFor: [
            {
              event: "start",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "start",
          priority: 0,
          request: [
            {
              detail: {
                value: "start",
              },
              event: "start",
            },
          ],
          waitFor: [
            {
              cb: [Function: cb],
              event: "",
            },
          ],
        },
      ],
      selectedEvent: {
        detail: {
          value: "start",
        },
        event: "start",
        priority: 0,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addHot",
          priority: 1,
          request: [
            {
              event: "hot",
            },
          ],
        },
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "cold",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "hot",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "hot",
        priority: 1,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      bThreads: [
        {
          name: "addCold",
          priority: 2,
          request: [
            {
              event: "cold",
            },
          ],
        },
        {
          block: [
            {
              event: "hot",
            },
          ],
          name: "mixHotCold",
          priority: 3,
          waitFor: [
            {
              event: "cold",
            },
          ],
        },
      ],
      selectedEvent: {
        event: "cold",
        priority: 2,
      },
    },
    type: "state-snapshot",
  },
  {
    data: {
      strategy: "chaos",
    },
    type: "end",
  },
]
`;
